AWS admin sign-up automation

nataliejschultz opened the pull request:
PR for e-mission/e-mission-docs#1008

email-config.py currently:

- Passes AWS credentials as environment variables
- Extracts email addresses from configuration files
- Checks for an AWS user pool associated with the configuration file (and can generate a new pool if this functionality is wanted)
- Checks for users in the pool by email address

will:
- Create accounts for users not in pool
- Send a customized (based on config file details) welcome email (triggered by addition of user to pool) with MFA instructions and password reset request

nataliejschultz commented:
First draft of the script is working and ready for review :)

nataliejschultz commented:
@shankari I realized you're not mentioned here, so probably didn't get my update! The script is ready for review 😄

shankari commented on email_automation/email-config.py:
```diff
@@ -0,0 +1,171 @@
+import boto3
+from botocore.exceptions import ClientError
```
These modules are not in the e-mission environment, or in this repo.
The script will not run as-is.
You need to have setup instructions and/or a setup script.

nataliejschultz replied:
I've written a github actions workflow file (since we'll have to have one anyway) and included this section in the steps:
```
- name: Install Boto3
    run: pip install boto3
```
I'll commit it after we discuss AWS authentication in our meeting tomorow, since I will likely have to add a section to support the GitHub OIDC provider.

shankari commented on email_automation/welcome-template.txt:
```diff
@@ -0,0 +1,34 @@
+    <br>
+        The login to the admin dashboard uses 2-factor authentication. You will have to install an authenticator app (e.g. Microsoft Authenticator, Google Authenticator) to receive One Time Passcodes (OTP) every time you want to log in. This is a requirement from NREL cyber to ensure that the data access to individual user information is secure.<br>
+      <br>
+    You can download trip and user tables as csv and view aggregate origin-destination points in the maps.<br>
```
this needs to be based on the config. If the `map_trip_lines` is enabled, then they can also view individual user origin-destination points.

nataliejschultz replied:
Updated this. See example of how it works [here](https://github.com/e-mission/e-mission-docs/issues/1008#issuecomment-1830736893)

shankari commented on email_automation/welcome-template.txt:
```diff
@@ -0,0 +1,34 @@
+      <br>
+    You can download trip and user tables as csv and view aggregate origin-destination points in the maps.<br>
+    <br>
+        Since you indicated that you want to map the data to infrastructure updates, your configuration includes trip start/end in the trip table. Let us know if you would like to exclude those for greater anonymity.
```
Ditto. This should be based on whether there are `"data_trips_columns_exclude"` columns or not.
If we do exclude the trip start/end, then they are already excluded, and this text should be changed.

nataliejschultz replied:
> Updated this. See example of how it works [here](https://github.com/e-mission/e-mission-docs/issues/1008#issuecomment-1830736893)

see above

shankari commented on email_automation/welcome-template.txt:
```diff
@@ -0,0 +1,34 @@
+    You can download trip and user tables as csv and view aggregate origin-destination points in the maps.<br>
+    <br>
+        Since you indicated that you want to map the data to infrastructure updates, your configuration includes trip start/end in the trip table. Let us know if you would like to exclude those for greater anonymity.
+    We will also support a trajectory table (<a href='https://github.com/e-mission/op-admin-dashboard/issues/55'>
```
In the months since this issue has been filed, the trajectory table has been enabled.

nataliejschultz replied:
Changed!

shankari commented on email_automation/email-config.py:
```diff
@@ -0,0 +1,171 @@
+logger = logging.getLogger()
+logger.setLevel(logging.INFO)
+
+#Input the filename as an argument in command line 
+filename_raw = sys.argv[1]
+filename = filename_raw.split(".")[0]
+pool_name = "nrelopenpath-prod-" + filename
+
```
nit: It is generally not considered good practice to write raw code like this, especially for files that are > ~ 50 lines long. You should use a `__main__ function instead.

nataliejschultz replied:
Added `if __name__ == "__main__":` function

shankari commented on email_automation/email-config.py:
```diff
@@ -0,0 +1,171 @@
+#Set up AWS credentials as environment variables + set variables 
+ACCESS = os.environ.get("AWS_ACCESS_KEY_ID")
+SECRET = os.environ.get("AWS_SECRET_ACCESS_KEY")
+TOKEN = os.environ.get("AWS_SESSION_TOKEN")
```
Where will we get these keys from? I can see that the KEY_ID and the ACCESS_KEY are likely static, but will users need to obtain a SESSION_TOKEN every time the script is run?

nataliejschultz replied:
I have set the variables by running the command line access exports (on the AWS start page) in my terminal, so other users can do the same since it is a copy-paste operation. The access keys expire after a certain amount of time, so they would need to be re-exported on most executions of this script.

In order to have the script work automatically with GitHub actions, we'll have to create an OIDC provider for GitHub + add an IAM role as detailed here. https://aws.amazon.com/blogs/security/use-iam-roles-to-connect-github-actions-to-actions-in-aws/

shankari commented on email_automation/email-config.py:
```diff
@@ -0,0 +1,171 @@
+    aws_session_token=TOKEN, 
+    region_name=AWS_REGION
+    )
+sts_client = boto3.client("sts")
```
does this not need any authentication?

shankari commented on email_automation/email-config.py:
```diff
@@ -0,0 +1,171 @@
+    UserPoolExist = False
+    #########One option to set the user pool without breaking (but still stop when condition met)
+    i = 0
+    while response["UserPools"][i]["Name"] != pool_name and i < len(response["UserPools"]) - 1:
+        print("looking for user pool...")
+        i = i + 1
+        if response["UserPools"][i]["Name"] == pool_name:
+            UserPoolExist = True
+            pool_id = response["UserPools"][i]["Id"]
+            print(pool_name + " pool exists! Checking for users...")
```
This is super clunky code. A better option would be:
1. map the responses to the name to get a name_list
2. check for existence in the name_array using python list methods (i.e. `in`)

shankari commented on email_automation/email-config.py:
```diff
@@ -0,0 +1,171 @@
+        print("looking for user pool...")
+        i = i + 1
+        if response["UserPools"][i]["Name"] == pool_name:
+            UserPoolExist = True
```
nit: you generally don't want to have variable names start with a capital - that looks like a class name. `is_userpool_exist` may be better.

nataliejschultz replied:
renamed exactly as you suggested!

shankari commented on email_automation/email-config.py:
```diff
@@ -0,0 +1,171 @@
+        response = cognito_client.list_users(UserPoolId=pool_id)
+        users = response["Users"]
+        result = False
+        for i in users:
+            for k, v in i.items():
+                if k == "Attributes":
+                    for j in v:
+                        if j["Name"] == "email":
+                            user_email = str(j["Value"])
+                            if str(email) == user_email:
+                                result = True
+        return result
```
Ditto wrt using list comprehensions instead of giant loops

nataliejschultz replied:
Rewrote the entire try block as such:
```
    try:
        response = cognito_client.list_users(UserPoolId=pool_id)
        users = response["Users"]
        result = False
        if str(users).find(email) > 1:
            result = True
        return result
```
find method will return a value of -1 if the string in question is not found.

shankari reviewed (changes requested):
@nataliejschultz I don't see any indicating of the testing done.
I understand that you might not have automated tests for this, but I don't even see any indication of manual testing done. In the future, please indicate testing done before sending it to me for review so I know that it is worth reviewing

Other changes below...

nataliejschultz commented on email_automation/email-config.py:
```diff
@@ -90,11 +76,11 @@ def email_extract():
-        # print("admindash", admindash_prefs)
-    return emails
+        admindash_prefs = data['admin_dashboard']
+        #maybe set a bool and if this is true, then add a section to the HTML. Otherwise leave it. Need to do the same for map trip lines tomorrow.
```
Removed

nataliejschultz commented on email_automation/email-config.py:
```diff
@@ -112,11 +98,21 @@ def create_account(pool_id, email, cognito_client):
-    return html
+        html = html.replace('<pool_name>', pool_name)
+        if map_trip_lines_enabled:
+            html = html.replace ('<map_trip_lines>', 'Additionally, you can view individual user-origin destination points.')
```
Updated!

nataliejschultz commented on email_automation/email-config.py:
```diff
@@ -112,11 +98,21 @@ def create_account(pool_id, email, cognito_client):
+        else:
+            html = html.replace ('<map_trip_lines>', '')
+        if 'data.start_loc.coordinates' in columns_exclude or 'data.end_loc.coordinates' in columns_exclude:
+            html = html.replace ('<columns_exclude>', 'Your configuration excludes trip start/end in the trip table. Let us know if you would like to include those.')
```
Updated!

nataliejschultz commented on email_automation/welcome-template.txt:
```diff
@@ -18,11 +18,12 @@
         The login to the admin dashboard uses 2-factor authentication. You will have to install an authenticator app (e.g. Microsoft Authenticator, Google Authenticator) to receive One Time Passcodes (OTP) every time you want to log in. This is a requirement from NREL cyber to ensure that the data access to individual user information is secure.<br>
       <br>
-    You can download trip and user tables as csv and view aggregate origin-destination points in the maps.<br>
+    You can download trip and user tables as a csv and view aggregate origin-destination points in the maps. <map_trip_lines> <br>
```
Updated

nataliejschultz commented on email_automation/welcome-template.txt:
```diff
@@ -18,11 +18,12 @@
+    <columns_exclude> 
+        
+    We also support a trajectory table.
+    https://github.com/e-mission/op-admin-dashboard/issues/55</a>)<br>
```
Removed

shankari commented on README.md:
```diff
@@ -10,6 +10,18 @@ Automation has been introduced for creating new config files. New deployers can
 
+Config file submission will begin the process of creating an admin dashboard and associated user pool for the project. In order to log in, users will need to generate an account. This process will be automated, but may occasionally need to be run manually if there are errors in the initial config submission. To generate an account:
+
+#Run the following command in terminal if boto3 is not installed
```
**future cleanup**: in the next PR, please indicate that they can install this in a virtualenv or a conda env to avoid polluting their main install.

I will not hold up the release for this, though. Please indicate that you have seen this using 👀 and we can resolve.

nataliejschultz replied:
Done 😃

shankari commented on README.md:
```diff
@@ -10,6 +10,18 @@ Automation has been introduced for creating new config files. New deployers can
+
+#Run the email-config.py script, and pass the name of the file in as an argument:
+
+`python email-config.py configfile.nrel-op.json`
```
**future cleanup**: Support a path to the filename. That is more unix-like and allows us to test out temporary files in non-standard locations as well. 
(the `os.path` module should help you parse out the filename from the path).

nataliejschultz replied:
I am not sure what you mean by this. Support a path as an argument for running in command line?

shankari replied:
Right now, we specify the filename - e.g. ``python email-config.py wyoming.nrel-op.json` and then you assume that the file location is `../config/wyoming.nrel-op.json`. 
https://github.com/e-mission/nrel-openpath-deploy-configs/pull/45/files#diff-406c93e4576589a3f4a4084699b1f01bf7c19a7b5cf7a3905561c2634431a499R18

I'm asking you to allow the user to specify the full filename (so I can have a copy of wyoming as `/tmp/wyoming-test-me-me-me.json` and use it instead.

The GH action would then invoke the script with `../config/<filename>`.

shankari commented on email_automation/welcome-template.txt:
```diff
@@ -6,8 +6,8 @@
-    <a href='https://<pool_name>-openpath.nrel.gov/admin/'>
-    https://<pool_name>-openpath.nrel.gov/admin/</a>.<br>
+    <a href='https://nrelopenpath-prod-<pool_name>-openpath.nrel.gov/admin/'>
+    https://nrelopenpath-prod-<pool_name>-openpath.nrel.gov/admin/</a>.<br>
```
I think this was supposed to fix https://github.com/e-mission/e-mission-docs/issues/1008#issuecomment-1839210067 but it is off. (sorry, partially my bad - I typed the original comment out in a hurry and edited it later).
This makes the URL even more wrong.

You should definitely test with other configs though and make sure that the link is valid.

nataliejschultz replied:
>You should definitely test with other configs though and make sure that the link is valid.

Are there any specific ones that you're okay with me adding myself to? I don't want to bother the program admins if I can avoid it, and I'm not sure if they get any emails when someone is added to their pools.

nataliejschultz replied:
> Make sure that you can actually login using the link in the email!

I've tested many times with both --local and --github, and realized that the variable `welcome` needed to be the same for both local and GitHub runs to be able to access welcome-template.txt. I've pushed that change and it's now ready!

shankari commented on email_automation/email-config.py:
```diff
@@ -74,10 +74,8 @@ def get_verified_arn(sts_client):
-        emails = [i.strip() for i in intro['admin_users'].split(",")]
         admindash_prefs = data['admin_dashboard']
-        #maybe set a bool and if this is true, then add a section to the HTML. Otherwise leave it. Need to do the same for map trip lines tomorrow.
+        emails = [i.strip() for i in admindash_prefs['admin_access'].split(",")]
```
**future cleanup**: can you add a link to the discussion around the choice of this location for the admin users?

nataliejschultz replied:
> **future cleanup**: can you add a link to the discussion around the choice of this location for the admin users?
Sent on teams; here is how the form will look from [Abby's PR](https://github.com/e-mission/nrel-openpath-deploy-configs/pull/49#issuecomment-1843223668)

shankari reviewed (changes requested):
- One real cleanup (the URL for the admin).
Multiple **future cleanups**. I would not have held up the release just for these, but since I am sending the PR back anyway, please get them handled as well.

shankari commented:
I would like you to support both. It sounds like the main difference is in authenticating to AWS. So you would have
`auth_for_gh_actions` and `auth_for_local_run` and have the rest of the code be the same.
You would invoke `auth_for_gh_actions` or `auth_for_local_run` either by seeing which environmental variables were present, or by using `argparse` to have people pass in `-g` for gha and `-l` for local or something.

shankari commented on email_automation/email-config.py:
```diff
@@ -4,40 +4,57 @@
     current_path = os.path.dirname(__file__)
-    config_path = os.path.relpath('../configs/'+ filename_raw, current_path)
+    maindir = current_path.rsplit("/",1)[0]
+    config_path = os.path.relpath('../configs/'+ filename_raw, current_path) if args.local else maindir + f'/configs/{filename_raw}'
```
@nataliejschultz if `args.local`, then the user can specify the full path to the file. I don't see how prepending `.../configs` will work here.

I think this should be:

```suggestion
    config_path = filepath_raw if args.local else maindir + f'/configs/{filename_raw}'
```

nataliejschultz replied:
Committed !

shankari commented on email_automation/email-config.py:
```diff
@@ -4,40 +4,57 @@
+    SECRET = os.environ.get("AWS_SECRET_ACCESS_KEY")
+    TOKEN = os.environ.get("AWS_SESSION_TOKEN")
+    AWS_REGION = "us-west-2"
+    welcome = 'welcome-template.txt'
```
where is this file? I don't see a `welcome.txt` in the PR here and this seems to imply that it is in the same directory as the script...

nataliejschultz replied:
See [comment](https://github.com/e-mission/nrel-openpath-deploy-configs/pull/45#discussion_r1433335001)

nataliejschultz replied:
The variable `welcome` changes between -l and -g, if that's what you're referencing!

shankari commented on email_automation/email-config.py:
```diff
@@ -116,12 +137,6 @@ def update_user_pool(pool_id, pool_name, html, identity_arn, cognito_client):
-        DeviceConfiguration={
-            'ChallengeRequiredOnNewDevice': True,
-            'DeviceOnlyRememberedOnUserPrompt': True
-        },
```
why is this removed?

nataliejschultz replied:
> why is this removed?

Jianli specifically asked me to remove it when we were going over the specifics of my usage of update_user_pool. These settings are already set the exact same way when the user pool is created, so there's no reason to add code to change it.

shankari reviewed (changes requested):
Almost there!

nataliejschultz commented on email_automation/welcome-template.txt:
In reference to[ this](https://github.com/e-mission/nrel-openpath-deploy-configs/pull/45/files#r1433259079) @shankari 

Is the file not showing up on your end? I can see it under Files Changed here

shankari replied:
Ah, I didn't see it because I was only looking at changes since the last time I viewed.
And there were no changes to the welcome file.

shankari commented:
@nataliejschultz This looks good to me overall and I am ready to merge it after the task below.
Can you please update with "Testing done" and move it to "Ready for review" once done?

nataliejschultz commented:
> @nataliejschultz Can you please update with "Testing done" (note [#45 (comment)](https://github.com/e-mission/nrel-openpath-deploy-configs/pull/45#discussion_r1417739879)) and move it to "Ready for review" once done?

I found an error with running the script from different directories while testing. Testing is now done, the script is updated, and the PR is back in ready for review! I think it's all ready.

shankari commented:
@nataliejschultz wrt "Can you please update with "Testing done"" I would like you to actually list the testing done.
e.g. 

<details><summary>.Testing from the same directory</summary>

```
```

</details>

...

nataliejschultz commented:
> I would like you to actually list the testing done. 

I removed myself from the account between each test to see if it would get through the entire process:

running script directly:
```
python email-config.py -l ../configs/wyoming.nrel-op.json
myemail@nrel.gov not in user pool! Creating account...
Account created! Sending welcome email.
```

running from one directory up: 

```
 python ./email_automation/email-config.py -l configs/wyoming.nrel-op.json 
myemail@nrel.gov not in user pool! Creating account...
Account created! Sending welcome email.
```

running at config file location:
```
python ../email_automation/email-config.py -l wyoming.nrel-op.json 
myemail@nrel.gov not in user pool! Creating account...
Account created! Sending welcome email.
```

running from far away directory:
```
python Documents/GitHub/forked-nrel-openpath-deploy-configs/email_automation/email-config.py -l Documents/GitHub/forked-nrel-openpath-deploy-configs/configs/wyoming.nrel-op.json 
myemail@nrel.gov not in user pool! Creating account...
Account created! Sending welcome email.
```
I got an email each time!

nataliejschultz commented:
I added functionality to remove users who are not in the config file, but are already in the user pool. Previously, I created the `user_already_exists` function:

```
def user_already_exists(pool_id, email, cognito_client):
    try:
        response = cognito_client.list_users(UserPoolId=pool_id)
        users = response["Users"]
        result = False
        if str(users).find(email) > 1:
            result = True
        return result
    except ClientError as err:
        logger.error(
            "Couldn't list users for %s. Here's why: %s: %s",
            pool_id,
            err.response["Error"]["Code"],
            err.response["Error"]["Message"],
        )
        raise
```
this function takes `users`, which is a list of dictionaries containing info on each user in the pool, and searches for a string matching the desired email. 

For the new use case, I modified it a bit to create `get_users`:

```
def get_users(pool_id, cognito_client):
    try:
        response = cognito_client.list_users(UserPoolId=pool_id)
        return response["Users"]
    except ClientError as err:
        logger.error(
            "Couldn't list users for %s. Here's why: %s: %s",
            pool_id,
            err.response["Error"]["Code"],
            err.response["Error"]["Message"],
        )
        raise
```
This function is called early on in the program, and gives us the `users` variable:
`users = get_users(pool_id, cognito_client)`

Next, each dictionary in `users` is looped over, and the email is isolated as `user_email`. 
```
    for user in users:
        for attr_dict in user["Attributes"]:
            if attr_dict["Name"] == "email":
                user_email = attr_dict["Value"]
                if user_email not in emails:
                    remove_user(pool_id, user_email)
                    print(f"{user_email} removed from pool.")
```
If the `user_email` is not in the `emails` list that was pulled from the config file, another new function named `remove_user` is called:


```
def remove_user(pool_id, user):
    response = cognito_client.admin_delete_user(
        UserPoolId= pool_id,
        Username= str(user)
)
```

I tested this by adding myself to the pool:
```
email1@nrel.gov not in user pool! Creating account...
Account created! Sending welcome email.
```

Then trying to re-add myself AND another new email,
```
email1@nrel.gov already in user pool!
email2@yahoo.com not in user pool! Creating account...
Account created! Sending welcome email.
```

And finally, removing my first email from the config to see what happens:

```
email1@nrel.gov removed from pool.
email2@yahoo.com already in user pool!
```
I checked the user pool, and my email1@nrel.gov was indeed removed from the pool! I decided to try and follow the welcome email to log in with my removed email, just to see what would happen. I got the following message:
<img width="466" alt="Screenshot 2024-01-31 at 11 20 25 PM" src="https://github.com/e-mission/nrel-openpath-deploy-configs/assets/90212258/eb59ad04-ab4a-46f7-ac86-a1ed24bb24b5">

nataliejschultz commented:
> I added functionality to remove users who are not in the config file, but are already in the user pool.

@shankari I haven't pushed these changes yet since this PR is still open; I am happy to put it in a separate PR once this one is merged, or just push it to this one. Wanted to check with you beforehand.
